charms/istio-pilot/src/charm.py

```diff
@@ -256,62 +217,35 @@ def handle_ingress_auth(self, event):
             for r in auth_routes
         )
 
-        self._delete_existing_resource_objects(
-            self.envoy_filter_resource, namespace=self.model.name
+        self._resource_handler.delete_existing_resource_objects(
+            self._custom_resources['envoy_filter'], namespace=self.model.name
         )
-        self._apply_manifest(auth_filters, namespace=self.model.name)
+        self._resource_handler.apply_manifest(auth_filters, namespace=self.model.name)
 
-    def _delete_object(
-        self, obj, namespace=None, ignore_not_found=False, ignore_unauthorized=False
-    ):
-        try:
-            self.lightkube_client.delete(type(obj), obj.metadata.name, namespace=namespace)
-        except ApiError as err:
-            self.log.exception("ApiError encountered while attempting to delete resource.")
-            if err.status.message is not None:
-                if "not found" in err.status.message and ignore_not_found:
-                    self.log.error(f"Ignoring not found error:\n{err.status.message}")
-                elif "(Unauthorized)" in err.status.message and ignore_unauthorized:
-                    # Ignore error from https://bugs.launchpad.net/juju/+bug/1941655
-                    self.log.error(f"Ignoring unauthorized error:\n{err.status.message}")
-                else:
-                    self.log.error(err.status.message)
-                    raise
-            else:
-                raise
-
-    def _delete_existing_resource_objects(
-        self,
-        resource,
-        namespace=None,
-        ignore_not_found=False,
-        ignore_unauthorized=False,
-        labels={},
-    ):
-        for obj in self.lightkube_client.list(
-            resource, labels={"app.juju.is/created-by": f"{self.app.name}"}.update(labels)
-        ):
-            self._delete_object(
-                obj,
-                namespace=namespace,
-                ignore_not_found=ignore_not_found,
-                ignore_unauthorized=ignore_unauthorized,
-            )
+    @property
+    def _custom_resources(self) -> dict:
```

Comment:
I don't think it really matters because this likely isn't called really frequently, but in general I think it would be good practice to avoid (re-)creating all resources whenever the user asks for any generic resource class.  It looks like Lightkube will pass you the existing resource if you've already made the same thing once before so that's good, but we still create extras on most passes (if I want virtual service, I still create envoy filter and gateway) and if I need another virtual service later I still read all the yamls and try to create things a second time

A possible approaches could be:
* using a getter rather than a normal property, creating whatever people ask as they're needed remembering the things you've already created and just returning them, something like:
```python
def __init__(self):
  ...
  self._custom_resource_classes = {}
def get_custom_resource_class(self, name: str) -> Resource:
  if name in self._custom_resource_classes:
    return self._custom_resource_classes[name]
  else:
    # Create the resource class
    ...
    self._custom_resource_classes[name] = new_resource_class
    return self._custom_resource_classes[name]
```

Reply:
I agree.